File: server/src/repositories/neo4j/character.repo.ts
import { getSession } from './connection';
import { Character, CharacterProfile, DisambiguationStatus } from '../../types';
import { v4 as uuid } from 'uuid';
import { getLogger } from '../../utils/logger';

const logger = getLogger();

function cosineSimilarity(a: number[], b: number[]): number {
  if (a.length !== b.length) return 0;
  let dot = 0, normA = 0, normB = 0;
  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i];
    normA += a[i] * a[i];
    normB += b[i] * b[i];
  }
  if (normA === 0 || normB === 0) return 0;
  return dot / (Math.sqrt(normA) * Math.sqrt(normB));
}

export class CharacterRepo {
  async create(data: Omit<Character, 'id'>): Promise<Character> {
    const session = getSession();
    try {
      const character: Character = { id: uuid(), ...data };
      await session.run(`CREATE (c:Character $props)`, { props: character });
      await session.run(
        `MATCH (n:Novel {id: $novelId}), (c:Character {id: $charId})
         CREATE (n)-[:HAS_CHARACTER]->(c)`,
        { novelId: data.novelId, charId: character.id }
      );
      return character;
    } finally {
      await session.close();
    }
  }

  async findById(id: string): Promise<Character | null> {
    const session = getSession();
    try {
      const result = await session.run(`MATCH (c:Character {id: $id}) RETURN c`, { id });
      if (result.records.length === 0) return null;
      return result.records[0].get('c').properties as Character;
    } finally {
      await session.close();
    }
  }

  async findByNovelId(novelId: string): Promise<Character[]> {
    const session = getSession();
    try {
      const result = await session.run(
        `MATCH (n:Novel {id: $novelId})-[:HAS_CHARACTER]->(c:Character) RETURN c`,
        { novelId }
      );
      return result.records.map(r => r.get('c').properties as Character);
    } finally {
      await session.close();
    }
  }

  async findByNameOrAlias(novelId: string, name: string): Promise<Character[]> {
    const session = getSession();
    try {
      const result = await session.run(
        `MATCH (n:Novel {id: $novelId})-[:HAS_CHARACTER]->(c:Character)
         WHERE c.name CONTAINS $name OR ANY(alias IN c.aliases WHERE alias CONTAINS $name)
         RETURN c`,
        { novelId, name }
      );
      return result.records.map(r => r.get('c').properties as Character);
    } finally {
      await session.close();
    }
  }

  async update(id: string, data: Partial<Character>): Promise<void> {
    const session = getSession();
    try {
      // 白名单校验：只允许更新已知字段
      const allowedFields = new Set([
        'name', 'aliases', 'gender', 'faction', 'identity', 'description',
        'firstAppearChapter', 'isProtagonist', 'protagonistOrder',
        'disambiguationStatus', 'novelId',
      ]);
      const safeData: Record<string, any> = {};
      for (const [k, v] of Object.entries(data)) {
        if (allowedFields.has(k)) safeData[k] = v;
      }
      if (Object.keys(safeData).length === 0) return;

      const setClauses = Object.keys(safeData).map(k => `c.${k} = $${k}`).join(', ');
      await session.run(
        `MATCH (c:Character {id: $id}) SET ${setClauses}`,
        { id, ...safeData }
      );
    } finally {
      await session.close();
    }
  }

  async mergeCharacters(primaryId: string, mergeIds: string[]): Promise<void> {
    const session = getSession();
    try {
      for (const mergeId of mergeIds) {
        // 1. 获取被合并角色的所有出边属性
        const outEdges = await session.run(
          `MATCH (merge:Character {id: $mergeId})-[r:RELATES_TO]->(c2:Character)
           RETURN c2.id AS targetId, properties(r) AS props`,
          { mergeId }
        );
        // 2. 获取被合并角色的所有入边属性
        const inEdges = await session.run(
          `MATCH (c2:Character)-[r:RELATES_TO]->(merge:Character {id: $mergeId})
           RETURN c2.id AS sourceId, properties(r) AS props`,
          { mergeId }
        );

        // 3. 为主角色创建新边（携带原属性）
        for (const record of outEdges.records) {
          const targetId = record.get('targetId');
          const props = record.get('props');
          await session.run(
            `MATCH (primary:Character {id: $primaryId}), (target:Character {id: $targetId})
             CREATE (primary)-[:RELATES_TO]->(target)`,
            { primaryId, targetId }
          );
          // 设置边属性
          if (props && Object.keys(props).length > 0) {
            const setClauses = Object.keys(props).map(k => `r.${k} = $prop_${k}`).join(', ');
            await session.run(
              `MATCH (primary:Character {id: $primaryId})-[r:RELATES_TO]->(target:Character {id: $targetId})
               SET ${setClauses}`,
              { primaryId, targetId, ...Object.fromEntries(Object.keys(props).map(k => [`prop_${k}`, props[k]])) }
            );
          }
        }
        for (const record of inEdges.records) {
          const sourceId = record.get('sourceId');
          const props = record.get('props');
          await session.run(
            `MATCH (source:Character {id: $sourceId}), (primary:Character {id: $primaryId})
             CREATE (source)-[:RELATES_TO]->(primary)`,
            { primaryId, sourceId }
          );
          if (props && Object.keys(props).length > 0) {
            const setClauses = Object.keys(props).map(k => `r.${k} = $prop_${k}`).join(', ');
            await session.run(
              `MATCH (source:Character {id: $sourceId})-[r:RELATES_TO]->(primary:Character {id: $primaryId})
               SET ${setClauses}`,
              { primaryId, sourceId, ...Object.fromEntries(Object.keys(props).map(k => [`prop_${k}`, props[k]])) }
            );
          }
        }

        // 4. 将被合并角色的名字添加到主角色别名
        await session.run(
          `MATCH (primary:Character {id: $primaryId}), (merge:Character {id: $mergeId})
           SET primary.aliases = primary.aliases + merge.name`,
          { primaryId, mergeId }
        );

        // 5. 删除被合并角色（DETACH DELETE 自动删除其所有关系）
        await session.run(`MATCH (c:Character {id: $mergeId}) DETACH DELETE c`, { mergeId });
      }
    } finally {
      await session.close();
    }
  }

  async search(novelId: string, keyword: string): Promise<Character[]> {
    const session = getSession();
    try {
      // 使用 CONTAINS 替代正则，避免注入风险
      // 同时搜索名字和别名
      const result = await session.run(
        `MATCH (n:Novel {id: $novelId})-[:HAS_CHARACTER]->(c:Character)
         WHERE c.name CONTAINS $keyword OR ANY(alias IN c.aliases WHERE alias CONTAINS $keyword)
         RETURN c`,
        { novelId, keyword }
      );
      return result.records.map(r => r.get('c').properties as Character);
    } finally {
      await session.close();
    }
  }

  async setProtagonist(id: string, isProtagonist: boolean, order?: number): Promise<void> {
    const session = getSession();
    try {
      const params: any = { id, isProtagonist };
      let extraSet = '';
      if (order !== undefined) {
        params.protagonistOrder = order;
        extraSet = ', c.protagonistOrder = $protagonistOrder';
      }
      await session.run(
        `MATCH (c:Character {id: $id}) SET c.isProtagonist = $isProtagonist${extraSet}`,
        params
      );
    } finally {
      await session.close();
    }
  }

  async getProtagonists(novelId: string): Promise<Character[]> {
    const session = getSession();
    try {
      const result = await session.run(
        `MATCH (n:Novel {id: $novelId})-[:HAS_CHARACTER]->(c:Character)
         WHERE c.isProtagonist = true
         RETURN c ORDER BY c.protagonistOrder`,
        { novelId }
      );
      return result.records.map(r => r.get('c').properties as Character);
    } finally {
      await session.close();
    }
  }

  /**
   * 为角色设置 embedding 向量
   */
  async setEmbedding(id: string, embedding: number[]): Promise<void> {
    const session = getSession();
    try {
      await session.run(
        `MATCH (c:Character {id: $id}) SET c.embedding = $embedding`,
        { id, embedding }
      );
    } finally {
      await session.close();
    }
  }

  /**
   * 向量相似度搜索（使用 Neo4j Vector Index）
   * 需要先创建向量索引
   */
  async vectorSearch(novelId: string, queryEmbedding: number[], topK: number = 10): Promise<Array<{ id: string; name: string; score: number }>> {
    const session = getSession();
    try {
      const result = await session.run(
        `MATCH (n:Novel {id: $novelId})-[:HAS_CHARACTER]->(c:Character)
         WHERE c.embedding IS NOT NULL
         CALL db.index.vector.queryNodes('character_embedding', $topK, $queryEmbedding)
         YIELD node, score
         WHERE node.id = c.id
         RETURN c.id AS id, c.name AS name, score`,
        { novelId, topK, queryEmbedding }
      );
      return result.records.map(r => ({
        id: r.get('id'),
        name: r.get('name'),
        score: r.get('score'),
      }));
    } catch (err) {
      // 向量索引可能不存在，回退到余弦相似度计算
      logger.warn(err, '向量索引查询失败，回退到手动计算');
      return this.fallbackVectorSearch(novelId, queryEmbedding, topK);
    } finally {
      await session.close();
    }
  }

  /**
   * 回退方案：手动计算余弦相似度
   */
  private async fallbackVectorSearch(novelId: string, queryEmbedding: number[], topK: number): Promise<Array<{ id: string; name: string; score: number }>> {
    const characters = await this.findByNovelId(novelId);
    const results: Array<{ id: string; name: string; score: number }> = [];

    for (const char of characters as any[]) {
      if (!char.embedding || !Array.isArray(char.embedding)) continue;
      const score = cosineSimilarity(queryEmbedding, char.embedding);
      results.push({ id: char.id, name: char.name, score });
    }

    return results.sort((a, b) => b.score - a.score).slice(0, topK);
  }

  /**
   * 创建向量索引（如果不存在）
   */
  async ensureVectorIndex(dimensions: number): Promise<void> {
    const session = getSession();
    try {
      // Neo4j 5.11+ 向量索引
      await session.run(
        `CREATE VECTOR INDEX character_embedding IF NOT EXISTS
         FOR (c:Character) ON (c.embedding)
         OPTIONS {indexConfig: {
           \`vector.dimensions\`: $dimensions,
           \`vector.similarity_function\`: 'cosine'
         }}`,
        { dimensions }
      );
    } catch (err) {
      logger.warn(err, '创建向量索引失败（可能 Neo4j 版本不支持或索引已存在）');
    } finally {
      await session.close();
    }
  }
}

export const characterRepo = new CharacterRepo();

File: services/api-gateway/internal/repository/neo4j/character_repo.go
package neo4j

import (
	"context"
	"fmt"
	"log/slog"
	"os"

	"github.com/yumeng001yu/ai-novel-character-graph/services/api-gateway/internal/model"

	"github.com/neo4j/neo4j-go-driver/v5/neo4j"
)

// CharacterRepo 角色数据仓库
type CharacterRepo struct{}

// NewCharacterRepo 创建角色仓库实例
func NewCharacterRepo() *CharacterRepo {
	return &CharacterRepo{}
}

// FindByID 根据 ID 查找角色
func (r *CharacterRepo) FindByID(ctx context.Context, id string) (*model.Character, error) {
	session := NewReadSession(ctx)
	defer session.Close(ctx)

	result, err := session.ExecuteRead(ctx, func(tx neo4j.ManagedTransaction) (interface{}, error) {
		cypher := `MATCH (c:Character {id: $id}) RETURN c`
		record, err := tx.Run(ctx, cypher, map[string]interface{}{"id": id})
		if err != nil {
			return nil, err
		}
		if record.Next(ctx) {
			return recordToCharacter(record.Record())
		}
		return nil, nil
	})
	if err != nil {
		slog.Error("查询角色失败", "id", id, "error", err)
		return nil, err
	}
	if result == nil {
		return nil, nil
	}
	return result.(*model.Character), nil
}

// FindByNovelId 根据小说 ID 查找所有角色
func (r *CharacterRepo) FindByNovelId(ctx context.Context, novelId string) ([]*model.Character, error) {
	// Neo4j 不可用时直接从文件系统读取
	if !IsAvailable() {
		return fsLoadCharacters(novelId), nil
	}

	session := NewReadSession(ctx)
	defer session.Close(ctx)

	result, err := session.ExecuteRead(ctx, func(tx neo4j.ManagedTransaction) (interface{}, error) {
		cypher := `
			MATCH (n:Novel {id: $novelId})-[:HAS_CHARACTER]->(c:Character)
			RETURN c ORDER BY c.firstAppearChapter`
		records, err := tx.Run(ctx, cypher, map[string]interface{}{"novelId": novelId})
		if err != nil {
			return nil, err
		}
		var characters []*model.Character
		for records.Next(ctx) {
			char, err := recordToCharacter(records.Record())
			if err != nil {
				return nil, err
			}
			characters = append(characters, char)
		}
		return characters, nil
	})
	if err != nil {
		// Neo4j 连接失败时返回空列表
		slog.Warn("查询小说角色失败（Neo4j 可能未启动），返回空列表", "novelId", novelId, "error", err)
		return []*model.Character{}, nil
	}
	if result == nil {
		return []*model.Character{}, nil
	}
	return result.([]*model.Character), nil
}

// FindByIds 根据多个 ID 查找角色
func (r *CharacterRepo) FindByIds(ctx context.Context, ids []string) ([]*model.Character, error) {
	session := NewReadSession(ctx)
	defer session.Close(ctx)

	result, err := session.ExecuteRead(ctx, func(tx neo4j.ManagedTransaction) (interface{}, error) {
		cypher := `MATCH (c:Character) WHERE c.id IN $ids RETURN c`
		records, err := tx.Run(ctx, cypher, map[string]interface{}{"ids": ids})
		if err != nil {
			return nil, err
		}
		var characters []*model.Character
		for records.Next(ctx) {
			char, err := recordToCharacter(records.Record())
			if err != nil {
				return nil, err
			}
			characters = append(characters, char)
		}
		return characters, nil
	})
	if err != nil {
		slog.Error("批量查询角色失败", "ids", ids, "error", err)
		return nil, err
	}
	if result == nil {
		return []*model.Character{}, nil
	}
	return result.([]*model.Character), nil
}

// Search 搜索角色（按名称或别名模糊匹配）
func (r *CharacterRepo) Search(ctx context.Context, novelId string, keyword string) ([]*model.Character, error) {
	session := NewReadSession(ctx)
	defer session.Close(ctx)

	result, err := session.ExecuteRead(ctx, func(tx neo4j.ManagedTransaction) (interface{}, error) {
		cypher := `
			MATCH (n:Novel {id: $novelId})-[:HAS_CHARACTER]->(c:Character)
			WHERE c.name CONTAINS $keyword OR ANY(alias IN c.aliases WHERE alias CONTAINS $keyword)
			RETURN c`
		records, err := tx.Run(ctx, cypher, map[string]interface{}{
			"novelId": novelId,
			"keyword": keyword,
		})
		if err != nil {
			return nil, err
		}
		var characters []*model.Character
		for records.Next(ctx) {
			char, err := recordToCharacter(records.Record())
			if err != nil {
				return nil, err
			}
			characters = append(characters, char)
		}
		return characters, nil
	})
	if err != nil {
		slog.Error("搜索角色失败", "novelId", novelId, "keyword", keyword, "error", err)
		return nil, err
	}
	if result == nil {
		return []*model.Character{}, nil
	}
	return result.([]*model.Character), nil
}

// Create 创建角色节点
func (r *CharacterRepo) Create(ctx context.Context, character *model.Character) error {
	// Neo4j 不可用时保存到文件系统
	if !IsAvailable() {
		fsAppendCharacter(character)
		return nil
	}

	session := NewSession(ctx)
	defer session.Close(ctx)

	_, err := session.ExecuteWrite(ctx, func(tx neo4j.ManagedTransaction) (interface{}, error) {
		cypher := `
			MATCH (n:Novel {id: $novelId})
			CREATE (c:Character {
				id: $id,
				novelId: $novelId,
				name: $name,
				aliases: $aliases,
				gender: $gender,
				faction: $faction,
				identity: $identity,
				personality: $personality,
				motivation: $motivation,
				firstAppearChapter: $firstAppearChapter,
				isProtagonist: $isProtagonist,
				disambiguationStatus: $disambiguationStatus
			})
			CREATE (n)-[:HAS_CHARACTER]->(c)
			RETURN c`
		_, err := tx.Run(ctx, cypher, map[string]interface{}{
			"id":                  character.ID,
			"novelId":             character.NovelID,
			"name":                character.Name,
			"aliases":             character.Aliases,
			"gender":              character.Gender,
			"faction":             character.Faction,
			"identity":            character.Identity,
			"personality":         character.Personality,
			"motivation":          character.Motivation,
			"firstAppearChapter":  character.FirstAppearChapter,
			"isProtagonist":       character.IsProtagonist,
			"disambiguationStatus": character.DisambiguationStatus,
		})
		return nil, err
	})
	if err != nil {
		slog.Error("创建角色失败", "error", err)
		return err
	}
	slog.Info("角色创建成功", "id", character.ID, "name", character.Name)
	return nil
}

// Update 更新角色信息
func (r *CharacterRepo) Update(ctx context.Context, character *model.Character) error {
	// Neo4j 不可用时更新文件系统
	if !IsAvailable() {
		fsAppendCharacter(character) // fsAppendCharacter 会按 ID 更新
		return nil
	}

	session := NewSession(ctx)
	defer session.Close(ctx)

	_, err := session.ExecuteWrite(ctx, func(tx neo4j.ManagedTransaction) (interface{}, error) {
		cypher := `
			MATCH (c:Character {id: $id})
			SET c.name = $name,
			    c.aliases = $aliases,
			    c.gender = $gender,
			    c.faction = $faction,
			    c.identity = $identity,
			    c.personality = $personality,
			    c.motivation = $motivation,
			    c.firstAppearChapter = $firstAppearChapter,
			    c.isProtagonist = $isProtagonist,
			    c.disambiguationStatus = $disambiguationStatus
			RETURN c`
		_, err := tx.Run(ctx, cypher, map[string]interface{}{
			"id":                  character.ID,
			"name":                character.Name,
			"aliases":             character.Aliases,
			"gender":              character.Gender,
			"faction":             character.Faction,
			"identity":            character.Identity,
			"personality":         character.Personality,
			"motivation":          character.Motivation,
			"firstAppearChapter":  character.FirstAppearChapter,
			"isProtagonist":       character.IsProtagonist,
			"disambiguationStatus": character.DisambiguationStatus,
		})
		return nil, err
	})
	if err != nil {
		slog.Error("更新角色失败", "id", character.ID, "error", err)
		return err
	}
	slog.Info("角色更新成功", "id", character.ID)
	return nil
}

// Delete 删除角色
func (r *CharacterRepo) Delete(ctx context.Context, id string) error {
	// Neo4j 不可用时从文件系统删除（需要遍历所有小说目录查找）
	if !IsAvailable() {
		// 简化处理：遍历所有小说目录，删除匹配的角色
		entries, err := os.ReadDir(fsNovelsDir)
		if err == nil {
			for _, entry := range entries {
				if !entry.IsDir() {
					continue
				}
				characters := fsLoadCharacters(entry.Name())
				for i, c := range characters {
					if c.ID == id {
						characters = append(characters[:i], characters[i+1:]...)
						fsSaveCharacters(entry.Name(), characters)
						return nil
					}
				}
			}
		}
		return nil
	}

	session := NewSession(ctx)
	defer session.Close(ctx)

	_, err := session.ExecuteWrite(ctx, func(tx neo4j.ManagedTransaction) (interface{}, error) {
		cypher := `MATCH (c:Character {id: $id}) DETACH DELETE c`
		_, err := tx.Run(ctx, cypher, map[string]interface{}{"id": id})
		return nil, err
	})
	if err != nil {
		slog.Error("删除角色失败", "id", id, "error", err)
		return err
	}
	slog.Info("角色删除成功", "id", id)
	return nil
}

// EnsureIndexes 确保必要的索引存在
func (r *CharacterRepo) EnsureIndexes(ctx context.Context) error {
	session := NewSession(ctx)
	defer session.Close(ctx)

	indexes := []string{
		"CREATE INDEX IF NOT EXISTS FOR (c:Character) ON (c.id)",
		"CREATE INDEX IF NOT EXISTS FOR (c:Character) ON (c.novelId)",
		"CREATE INDEX IF NOT EXISTS FOR (c:Character) ON (c.name)",
	}

	_, err := session.ExecuteWrite(ctx, func(tx neo4j.ManagedTransaction) (interface{}, error) {
		for _, indexCypher := range indexes {
			_, err := tx.Run(ctx, indexCypher, nil)
			if err != nil {
				return nil, fmt.Errorf("创建索引失败: %w", err)
			}
		}
		return nil, nil
	})
	if err != nil {
		slog.Error("确保角色索引失败", "error", err)
		return err
	}
	slog.Info("角色索引确保完成")
	return nil
}

// recordToCharacter 将 Neo4j 记录转换为 Character 模型
func recordToCharacter(record *neo4j.Record) (*model.Character, error) {
	node, ok := record.Get("c")
	if !ok {
		return nil, fmt.Errorf("记录中未找到节点 c")
	}
	n, ok := node.(neo4j.Node)
	if !ok {
		return nil, fmt.Errorf("节点类型断言失败")
	}
	props := n.Props

	// 处理 aliases 切片
	var aliases []string
	if v, ok := props["aliases"]; ok && v != nil {
		if arr, ok := v.([]interface{}); ok {
			for _, item := range arr {
				if s, ok := item.(string); ok {
					aliases = append(aliases, s)
				}
			}
		}
	}

	character := &model.Character{
		ID:                  getStringProp(props, "id"),
		NovelID:             getStringProp(props, "novelId"),
		Name:                getStringProp(props, "name"),
		Aliases:             aliases,
		Gender:              getStringProp(props, "gender"),
		Faction:             getStringProp(props, "faction"),
		Identity:            getStringProp(props, "identity"),
		Personality:         getStringProp(props, "personality"),
		Motivation:          getStringProp(props, "motivation"),
		FirstAppearChapter:  getIntProp(props, "firstAppearChapter"),
		IsProtagonist:       getBoolProp(props, "isProtagonist"),
		DisambiguationStatus: getStringProp(props, "disambiguationStatus"),
	}
	return character, nil
}
